File: src/301-remove-invalid-parentheses.py
# 301. Remove Invalid Parentheses


class Solution:
    def removeInvalidParentheses(self, s: str) -> list:
        left_num = 0  # max number of '(' to remove
        right_num = 0  # max number of ')' to remove
        for ch in s:
            if ch == '(':
                left_num += 1
            elif ch == ')':
                if left_num:
                    left_num -= 1
                else:
                    right_num += 1
        res = set()
        self.dfs(left_num, right_num, 0, '', 0, s, res)
        return list(res)

    def dfs(self, left_num, right_num, open, path, i, s, res):
        if left_num < 0 or right_num < 0 or open < 0:
            return
        elif i == len(s):
            if left_num == 0 and right_num == 0 and open == 0:
                res.add(path)
        else:
            ch = s[i]
            if ch == '(':
                # add
                self.dfs(left_num, right_num, open + 1, path + ch, i + 1, s, res)
                # skip
                self.dfs(left_num - 1, right_num, open, path, i + 1, s, res)
            elif ch == ')':
                # add
                self.dfs(left_num, right_num, open - 1, path + ch, i + 1, s, res)
                # skip
                self.dfs(left_num, right_num - 1, open, path, i + 1, s, res)
            else:
                # add s[i]
                self.dfs(left_num, right_num, open, path + ch, i + 1, s, res)


if __name__ == "__main__":
    import time
    start = time.clock()

    s = Solution()

    test_case = '()())()'

    ans = s.removeInvalidParentheses(test_case)

    print(ans)

    end = time.clock()
    print('Running time: %s ms' % ((end - start) * 1000))

File: src/301-remove-invalid-parentheses.cpp
// 301. Remove Invalid Parentheses


#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
	vector<string> removeInvalidParentheses(string s) {
		string path = "";
		unordered_set<string> result_set;
		int max_len = 0;
		dfs(0, 0, path, max_len, s, s.length(), result_set);
		vector<string> result(result_set.begin(), result_set.end());
		return result;
	}
private:
	void dfs(int start, int left_cnt, string& path, int& max_len, const string& full_str, const int full_len, unordered_set<string>& result_set) {
		if (start == full_len) {
			if (left_cnt == 0 && path.length() >= max_len) {
				if (path.length() > max_len) {
					result_set.clear();
					max_len = path.length();
				}
				result_set.insert(path);
			}
			return;
		}
		int ori_start = start;
		while (start < full_len && full_str[start] != '(' && full_str[start] != ')') {
			path += full_str[start];
			++start;
		}
		if (start != full_len) {
			dfs(start + 1, left_cnt, path, max_len, full_str, full_len, result_set);
			if (full_str[start] == '(') {
				path.push_back('(');
				dfs(start + 1, left_cnt + 1, path, max_len, full_str, full_len, result_set);
				path.pop_back();
			}
			else {
				if (left_cnt != 0) {
					path.push_back(')');
					dfs(start + 1, left_cnt - 1, path, max_len, full_str, full_len, result_set);
					path.pop_back();
				}
			}
		}
		else {
			dfs(start, left_cnt, path, max_len, full_str, full_len, result_set);
		}
		while (start > ori_start) {
			path.pop_back();
			--start;
		}

	}
};

//int main() {
//	Solution s = Solution();
//
//	string testcase = "()())()";
//	
//	vector<string> ans = s.removeInvalidParentheses(testcase);
//
//	for (const string& str : ans)
//		cout << "\"" << str << "\"" << endl;
//
//	system("pause");
//	return 0;
// }